### ISSUE TITLE
Check the gpg signature when updating

### ISSUE THREAD

**Issue by pamarcos** [AppImage/AppImageUpdate#16]
This issue is created just for reference so that this is not forgotten.

At the moment, the gpg_check method is commented out as you can see in the current implementation:

```
gpg_check()
{
  set +e
  if [ "$(which gpg_currently_disabled)" != "" ] ; then
    echo "GPG is installed; hence attempting signature verification"
    ASC_URL=$( echo "${ZSYNC_URL}" | sed -e 's|.zsync|.asc|g' )
    # curl -k -L -O -# "${ASC_URL}" && gpg --verify $(basename "${ASC_URL}") && echo "Signature verified"
  else
    echo "GPG is not installed; hence skipping signature verification" >&2
  fi
  set -e
}
```

After reading https://github.com/AppImage/AppImageKit/issues/238, I came up with the following solution:

```
gpg_check()
{
  set +e
  if [[ "$(which gpg2)" != "" && "$(which validate)" != "" ]]; then
    echo "GPG is installed; hence attempting signature verification"
    validate "${ISO}"

    if [[ $? -eq 0 ]]; then
      echo "GPG signature verified"
    else
      echo "GPG verification failed. Please close and retry"
      mv -f "${ISO}.zs-old" "${ISO}"
      exit 1
    fi
  else
    echo "GPG is not installed; hence skipping signature verification" >&2
  fi
  set -e
}
```

All that's needed is to include the `validate` binary into the PATH so that `appimageupdate` can run it. In my case, I created a self update-able AppImage adding those two plus `appimageupdategui` and `zsync_curl`.

**Comment by probonopd** [comment]
Thanks @pamarcos. Please be aware that @TheAssassin is currently rewriting AppImageUpdate in C++, so it's probably not worthwhile to continue working on the bash-based AppImageUpdate script.

**Comment by TheAssassin** [comment]
Taking this as a feature proposal for the rewrite.

**Comment by probonopd** [comment]
We should think about what happens in which case:
* File is not signed -> Warning but user can choose to continue
* File is signed but signature cannot be verified because public key is not in user's keychain -> Warning but user can choose to continue
* File is signed and signature can be verified using public key in user's keychain -> No warning
* Signature of updated file does not match signature of original file -> Warning but user can continue

Over time, we may opt to make "user can continue" more cumbersome in order to prevent users from just clicking "OK" all the time.

**Comment by TheAssassin** [comment]
You don't need the keychain at all, or should give it a lower priority. I plan to extract the signature from the previous AppImage, and then use its public key to verify any subsequent signature. This is quite secure, and doesn't bug the user too much or hack around with his private keychain. Speaking of keychains, we'd have to set up a separate one for these purposes anyway. And nothing really stops a process of putting in new keys into any keychain owned by the current user, so that doesn't really provide any security.

**Comment by probonopd** [comment]
> I plan to extract the signature from the previous AppImage, and then use its public key to verify any subsequent signature.

Where do you get the public key from, if not from the keychain?

**Comment by TheAssassin** [comment]
Hm, good point. They're not bundled with the signature (although that would make a `lot` of sense actually, they're public anyway).

What I want to say is, we should not alter the user's private GPG keyring. What we could do however is create our own keyring in e.g., `.config/appimage/update/keyring.gpg` or something like that.

Since I'd anyway try to fetch the missing key from a PGP server (like, `pgp.mit.edu`) if it's not available locally, we could ask the user to confirm the key, and eventually put it in AppImageUpdate's keyring. I know, most people are most likely just going to hit `OK` anyway, but privacy aware people at least have a chance to audit their keyring, which provides a core principle of secure systems, being transparent to the user.

**Comment by probonopd** [comment]
I don't think we should alter anything. How the user gets the public keys into his keyring is not something we should be concerned with, at least initially. I could imagine that e.g., openSUSE would pre-install their public keyring into the system's default keychain.

If as a user I trust e.g., krita.org, then _I as the user_ could import their key into my keyring if I want `appimaged` to set their AppImages as trustworthy (e.g., run outside of a confined Firejail sandbox).

At least initially.

But this is up for discussion.

**Comment by TheAssassin** [comment]
I started working on this feature. For now, I plan to show a dialog whenever a validation fails, to give the user feedback about what's going on (given there's a signature, if there is not, a warning will be logged). We could also just fail the updates, but I don't think that average users would understand why an update has failed.

The validation should work like this:
  - AppImageUpdate checks whether a signature is contained in the new file
  - A check is performed whether the signature's contents apply (i.e., the hash sum matches)
  - Now, there's a couple of options how this can continue:
    - signature found in the old AppImage (this works since for a short amount of time, we have both old and new file available for these purposes):
      - if the key is found in the local keychain(s), the signature will be validated using that key
      - if the key is not available in the local keychain, but the same key has been used for both signatures,  AppImageUpdate will attempt to download the key to a temporary keychain and validate the new signature
        - if that fails, the validation fails
    - if there is no signature in the old AppImage:
      - if the key is available in the local keychain, the signature will be validated using that key
      - if the key is not available in the local keychain, but from a key server, the user is prompted (with a "warning icon", as it's less critical than a failed validation) to confirm that this key shall be used for this specific update, then a temporary keychain is created and the key will be used for validation
      - otherwise, the validation will fail

*Note for myself: workflow graphs in bullet lists aren't very readable.*

Problems to solve:

  - How can this be implemented in a reusable way? Should the code go into libappimageupdate, or a GUI specific module?
    - Does libappimageupdate need an API redesign to allow it to post warnings and block the update process?
    - Should the verification be offered as a synchronous method of the Updater class, similar to the update check? This could, like the update check, be combined with the status message API.
  - Is it sufficient to show a dialog on failed validations? Should AppImageUpdate refuse to work entirely?
    - Do we need to recommend users of the update information and signatures to test updates before publishing releases? Do we have to write a tool for this?

**Comment by TheAssassin** [comment]
I pushed an API proposal in .

**Comment by probonopd** [comment]
We should think about the UI for this. Especially don't spew warnings all over the place, which will only do two things:
* Give the impression that AppImage is totally insecure (in general)
* Make users ignore all warnings

So what I could imagine is:
* Do not warn if an AppImage has no signature at all
* Give extra credit (think "green lock in the browser bar") for an AppImage that is signed and the signature of which can be verified using a trusted key in the keychain
* Show a warning dialog only if something nasty is going on (e.g., the new signature of an AppImage does not match the pre-update one, or the AppImage has no signature anymore post-update while it did have one pre-update, etc.)

**Comment by TheAssassin** [comment]
> the new signature of an AppImage does not match the pre-update one

Doesn't make sense. Please elaborate.

I value security way higher than any user experience, and took a web browser's SSL certificate verification as a kind of role model for designing my workflow here.

I don't see a problem with the UX I suggested. Especially when comparing to what e.g., browsers, do. A browser will show a warning if a certificate's origin can't be verified, and will show an error when the crypto doesn't work.

I basically suggested the same methods. I already mentioned if there's no signatures involved, I wouldn't show a warning or error at all. And if there are, and they work fine, it's all good. However, if the verification doesn't work, you should warn and/or fail, depending on the severity of the issue.

I mean, why bother verifying it at all, if you're not informing the user about it properly, right?

If you as an app author screw up the signature part, why should an update be performed at all? If it's just your key that is unknown, a simple confirmation dialog won't prevent anybody from performing the update (in web browser words: "add a temporary exception").
I'd add browser-like icons to the dialogs, so people who are familiar with the UX of a web browser will feel comfortable with AppImageUpdate, too.

But picking up your idea about the "extra security" credit in the progress bar: We could make it yellow if there's no signatures involved, and write something to the progress label like "Warning: the update has not been signed." But that's basically an extension to what I suggested above.

**Comment by probonopd** [comment]
>  took a web browser's SSL certificate verification as a kind of role model

Yes, I think thats the right approach.

But once Chrome starts giving _warnings_ on _every_ http site, I'd consider that overdone. For AppImages, we're still at a stage where the majority comes without the "secure" (signed) option. Don't throw warnings on those, at least not yet.

**Comment by TheAssassin** [comment]
> (given there's a signature, if there is not, a warning will be logged).

Where did you read that whole "no signature = show warning" thing? I never proposed such a thing. I suggested to _log_ a message (i.e., to the "Details" widget)...

**Comment by probonopd** [comment]
__Aaaah... ok___

**Comment by TheAssassin** [comment]
I noticed my API proposal won't really work with the workflow I suggested.

While calling `validateSignature()` would work the way it's implemented right now, one can't "interrupt" an update if a validation fails, as the update thread might have removed the former AppImage. To mitigate this problem (i.e., be able to restore the old AppImage), one would have to copy (or create a hardlink) the old AppImage before starting the `Updater`. That's some complexity I don't want to have in the API.

Instead, we could think about implementing a callback-based method.

**Comment by probonopd** [comment]
> the update thread might have removed the former AppImage

What do you mean by "removed"? I think we had established before that "updating" an application should mean that the newer version is downloaded but the older version is not (automatically; by default) deleted, so that users have a chance to go back to the old version if they don't like the new one.

**Comment by TheAssassin** [comment]
> What do you mean by "removed"? I think we had established before that "updating" an application should mean that the newer version is downloaded but the older version is not (automatically; by default) deleted, so that users have a chance to go back to the old version if they don't like the new one.

Right. But there's the flags `-r` and `-O` that might interfere with this workflow. Further investigation needed.

**Comment by TheAssassin** [comment]
I got the basic workflow implemented in appimageupdatetool. I decided to abort on errors, and show a message on warnings. I might extend the latter to prompt the user whether to continue, but I assume people would then start to ask for a `-y` or `--force-yes` flag or something. We'll have to discuss that later on, when the validation code has been verified.

A few problems I came across so far:

  - Working with gpg2 via the CLI is hard. We cannot link to GPGME (or libgpg), as they are licensed under the terms of the GPLv3. Therefore, I had to work around this by using gpg2 as a subprocess.
  - We need to discuss whether we should bundle a copy of gpg2, to make sure it's available. At the moment, the update process is aborted whenever gpg2 is missing.
  - @probonopd's "official" signing key used to sign AppImageKit etc. hasn't been published anywhere, neither keyservers nor otherwise (at least I couldn't find any pubkey file). That leads to validation errors when updating AppImageKit and/or appimaged.
  - As a consequence, we could bundle some "known good" keys in a keyring that we ship with AppImageUpdate. I think it's a good idea to do so, but I'd like to hear more opinions on that. I would start by bundling our key and the openSUSE build service keys (CC @adrianschroeter).

In order to be able to test the full workflow, we'll need a couple of test projects and (ideally) a lot of testers.

**Comment by probonopd** [comment]
> We need to discuss whether we should bundle a copy of gpg2, to make sure it's available. At the moment, the update process is aborted whenever gpg2 is missing.

For me, the the whole signature thing should be treated as an optional security benefit, with a fallback to not using it at all if it is not there. Users who want to take advantage of the gpg2 based signature stuff need to install gpg2. Users who don't want or need this should not be bothered with errors; but they should simply not see the equivalent of a "green lock symbol" either.

**Comment by probonopd** [comment]
> we could bundle some "known good" keys in a keyring

This is where the "web of trust" idea comes in, so that we don't have to decide which keys to include. To be honest, I'd rather like to see the keyring part in the distribution. They already today decide which keys to ship by default, and users are always free to add extra keys.

**Comment by TheAssassin** [comment]
"web of trust" doesn't work the way you need it to work, as mentioned more than a couple of times. It's like communism: nice concept, but in the real world...

Also, not an excuse for not publishing the public keys e.g., on the keyservers. If our key would have been on a keyserver, for example, the signature verification would work already fine.

**Comment by probonopd** [comment]
Only that keyservers are centralistic whereas I am imagining something more _web_ of trust like... as in p2p. But I understand that this seems not viable at this point in time.

In the meantime, is it reasonable to leave it to _distributions_ to decide which keys to ship by default? I mean, I would expect openSUSE to decide which of the keys used on OBS are to be trusted on which level..

**Comment by TheAssassin** [comment]
Okay, then let me get more explicit. Please upload the public key, either to the keyservers, or as a file, so I can use them!

**Comment by probonopd** [comment]
This one?
https://github.com/AppImage/AppImageKit/blob/appimagetool/master/resources/pubkey.asc

**Comment by TheAssassin** [comment]
Don't know how I could miss that file... I was looking for something with `key` (or `.key`?) in its name... But yes, I think so.

**Comment by adrianschroeter** [comment]
@probonopd from OBS POV, we could found an appimage project official OBS project, where developers can submit request their appimage builds. A team (you?) needs to review them and they would be built using that projects signing key. We could add this key then to the updater ....

**Comment by TheAssassin** [comment]
@adrianschroeter how are the files signed at the moment? They do contain a signature, as far as I can see...

**Comment by TheAssassin** [comment]
The OBS files are apparently signed with the home projects' keys. Therefore, we need a way to temporarily download the keys from OBS's systems in order to validate these signatures.

Another issue:
```
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.

**Comment by TheAssassin** [comment]
Okay, I tried with some external aria2 AppImage built on OBS, to see whether AppImageUpdate can update the file.

```
gpg2: gpg: Signature made Sun Oct 29 03:01:06 2017 CET using RSA key ID A3D71FBA
gpg2: gpg: Good signature from "home:nandub OBS Project <home:nandub@build.opensuse.org>" [unknown]
gpg2: gpg: WARNING: This key is not certified with a trusted signature!
gpg2: gpg:          There is no indication that the signature belongs to the owner.
gpg2: Primary key fingerprint: FAE5 2C87 F1A3 BA54 51FA  6942 C8CB A639 A3D7 1FBA
gpg2: gpg: Signature made Mon Dec  4 00:05:23 2017 CET using RSA key ID A3D71FBA
gpg2: gpg: BAD signature from "home:nandub OBS Project <home:nandub@build.opensuse.org>" [unknown]
Validation error: Bad signature
```

So apparently, the original AppImage's signature can be verified successfully (despite the warning mentioned in the last comment), but the verification of the updated AppImage didn't work. Investigating this at the moment.

**Comment by probonopd** [comment]
> @probonopd from OBS POV, we could found an appimage project official OBS project, where developers can submit request their appimage builds. A team (you?) needs to review them and they would be built using that projects signing key. We could add this key then to the updater ....

Hi @adrianschroeter do you see a chance that the AppImages coming from projects that are in one of the official distributions would be signed with the distribution's key? (These are probably not the correct OBS terms, so let me give an example: "The AppImage of FreeCAD that is in Leap would be signed by a key that Leap trusts")

Reason is that our team is very small and we don't have the capacity to decide which projects should and which should not be signed. At least I personally would follow the level of trust that the openSUSE project gives to a particular project and/or maintainer.

**Comment by TheAssassin** [comment]
@probonopd if they had a OBS keyserver so that the keys could at least be searched for, that'd be a great step ahead, and would suffice for me.

**Comment by adrianschroeter** [comment]
well, yes, if you would manage to get them into these projects, they would get signed with the distro key.

It might be just hard to convince the distro people to accept them there, because there is no direct benefit for them.

however, no matter which key is used, you could bless it with your official gpg key?

**Comment by TheAssassin** [comment]
I agree that using a distro(!) key for external software is a bad idea. Therefore I suggested to build a keyserver with all the OBS keys, so that they can be queried from the gpg2 CLI.

> however, no matter which key is used, you could bless it with your official gpg key?

Could you elaborate on that point? I don't understand what you're suggesting.

**Comment by adrianschroeter** [comment]
well, but what would you win with a key server? You would still need to verify the key IMHO.

So, you could ship the by default trusted keys with your source already, no?

**Comment by TheAssassin** [comment]
@adrianschroeter the trust model is "if the old AppImage is properly signed with the key, we trust it". So, what we'd need is the public key to actually verify the signatures.

But you're right, the whole issue can be resolved by shipping the actual pubkey with the AppImages. I have to ask myself why I didn't come up with that myself...

**Comment by probonopd** [comment]
> I agree that using a distro(!) key for external software is a bad idea. 

Just to be clear, I was not suggesting this. Instead, I was suggesting to sign AppImages of packages that are _inside_ the distribution already with the distribution key.

> But you're right, the whole issue can be resolved by shipping the actual pubkey with the AppImages.

How does this solve the question _which_ pubkeys to trust?

**Comment by TheAssassin** [comment]
You probably misunderstood me. That doesn't solve the question at all. And I never _tried_ to solve that question at the moment. The issue we are working on here is, you can't verify a signature's integrity without having the key. When you have the key, you can at least verify the signature applies. See https://github.com/AppImage/AppImageUpdate/issues/16#issuecomment-369498112.

Then, you can continue with the trust model workflow I suggested, which solves the question whether the key can be trusted or not. That's a completely different issue.

**Comment by probonopd** [comment]
OK. For the user, I conceptually would like to simplify it to three states:

* Normal (grey) = there is no signature or the signature is not trusted by the user (i.e., the pubkey is not trusted in the keyring)
* "Green lock" = the signature has been successfully verified against a pubkey that is trusted in the keyring (because the user or the distro has marked the key as trusted and has put it in the keyring)
* "RED FLAG" = the signature does not match the pre-update ("old file") signature, or has been tampered with in some other way

Does this make sense? Do we need more user-visible states? (What would "The signature has been successfully verified against an untrusted signature" mean to the "normal user"? Yellow? How does e.g., `apt-get` treat this state)?

**Comment by TheAssassin** [comment]
There is a need for a "less critical error" a.k.a. warning state. In this state, there is some irregularity, but it can be solved, e.g., by asking the user questions like "I don't know this key, shall we trust it?" and have the user verify the data. Of course it's not ideal to have a user check a PGP key's data (see https://en.wikipedia.org/wiki/Key_server_(cryptographic)#Problems_with_keyservers for more information), but it is not a fatal error either.

**Comment by TheAssassin** [comment]
Also, I don't want to introduce a "trusted keyring" at all. It makes everything overly complicated.

**Comment by TheAssassin** [comment]
I merged a PR in AppImageKit to embed the signatures' keys in the AppImages as well. I'll implement the code to extract the signatures tonight. And I'll drop the "trusted keyring" again, as especially we should make sure we properly maintain the AppImages' signatures.

**Comment by probonopd** [comment]
> We still get BAD signatures with our AppImages.

Would me having mixed up the public signature be an explanation for this? I remember it took me a couple of tries to a useable secret key, maybe I did not keep the public key in sync. I will have to check. This whole key handling stuff is a real pain.

**Comment by TheAssassin** [comment]
The public key is embedded now, and it's directly exported from the private key using the `gpg2` CLI. It's the same key as in the repo, as far as I can tell.

**Comment by probonopd** [comment]
> asking the user questions like "I don't know this key, shall we trust it?"

This is exactly the kind of user experience I would really want us to try hard to avoid. "Average Joe" will not understand the question, click "Yes", and think "this AppImage stuff is complicated".

**Comment by TheAssassin** [comment]
That's why I removed that yesterday. Please try the latest Qt UI.

**Comment by TheAssassin** [comment]
I am working on this again. We verified the pubkey in the repo is the one of the key we use to sign the appimagetool/appimaged AppImages at the moment. That means there must be an issue in the signature verification. I am quite sure it's a workflow problem. AppImageUpdate calculates the SHA checksum _with_ the signature key, but it seems appimagetool calculates its hash _before_ embedding the key.

**Comment by probonopd** [comment]
> AppImageUpdate calculates the SHA checksum with the signature key, but it seems appimagetool calculates its hash before embedding the key.

Since we use a digest that excludes (skips) the are in which the key is stored, this _should_ not matter.

**Comment by TheAssassin** [comment]
@probonopd it's a misconception in the implementation in AppImageUpdate that I found, which I'm going to fix soon (hopefully).
